### PR TITLE
Feature rank

### PR DESCRIPTION
**Pull request by blackbbc**

### CODE REVIEW

File: vj4/cron_jobs/rank.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
+from pymongo import MongoClient
```
Comment (iceboy-sjtu) [inline vj4/cron_jobs/rank.py]: Don't import classes (and functions). Import modules.

```pythonn
import pymongo
import pymongo.errors
```

File: vj4/cron_jobs/rank.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    result.append(rank)
+
+  return result
+
+def count_level(perc):
+  global level_config
+
+  perc *= 100
+
+  for item in level_config:
```
Comment (iceboy-sjtu) [inline vj4/cron_jobs/rank.py]:
```python
for key, value in level_config:
```

File: vj4/cron_jobs/rank.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    elif udoc['rp'] == stack[-1]['rp']:
+      stack.append(udoc)
+    else:
+      while len(stack) != 0:
+        doc = stack.pop()
+        result.append(rank)
+      rank += 1
+      stack.append(udoc)
+
+  while len(stack) != 0:
```
Comment (iceboy-sjtu) [inline vj4/cron_jobs/rank.py]:
```python
while stack:
```

File: vj4/cron_jobs/rank.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
+from pymongo import MongoClient
+from pymongo.errors import BulkWriteError
+
+from vj4.util import options
+
+options.define('db_host', default='localhost', help='Database hostname or IP address.')
+options.define('db_port', default=27017, help='Database port.')
+options.define('db_name', default='test', help='Database name.')
+options.define('coll_name', default='user', help='User collection')
```
Comment (iceboy-sjtu) [inline vj4/cron_jobs/rank.py]: `coll_name` should not be a option.

File: vj4/cron_jobs/rank.py [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+    value = item[1]
+    if perc <= value:
+      return key
+
+def handle_rank():
+  conn = MongoClient(options.options.db_host, options.options.db_port)
+  db = conn[options.options.db_name]
+  coll = db[options.options.coll_name]
+
+  udocs = list(coll.find())
```
Comment (iceboy-sjtu) [inline vj4/cron_jobs/rank.py]: You should not read all users into memory. Consider:

```python
udocs = coll.find().sort([('rp', -1)])
for udoc in udocs:
  ...
```

File: vj4/cron_jobs/rank.py [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+  udocs.sort(key = lambda x:(x['rp']), reverse=True)
+
+  rank_array = rank_data(udocs)
+  total = rank_array[-1]
+
+  index = 0
+  bulk = coll.initialize_unordered_bulk_op()
```
Comment (iceboy-sjtu) [inline vj4/cron_jobs/rank.py]: I've never seen this before. What are the benefits and implications?

File: vj4/cron_jobs/rank.py [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+
+  perc *= 100
+
+  for item in level_config:
+    key = item[0]
+    value = item[1]
+    if perc <= value:
+      return key
+
+def handle_rank():
```
Comment (iceboy-sjtu) [inline vj4/cron_jobs/rank.py]: This function should not be called `handle_xxx`

File: vj4/cron_jobs/rank.py [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+  udocs = list(coll.find())
+  udocs.sort(key = lambda x:(x['rp']), reverse=True)
+
+  rank_array = rank_data(udocs)
+  total = rank_array[-1]
+
+  index = 0
+  bulk = coll.initialize_unordered_bulk_op()
+
+  for udoc in udocs:
```
Comment (iceboy-sjtu) [inline vj4/cron_jobs/rank.py]: First, you should not use index for iterating over two sequences. You may:

```python
for udoc, rankN in zip(udocs, rank_array):
  ...
```

Second, you should not store all ranks in memory. Consider either make the rank method process one user at a time (with a state stored somewhere) or make it a generator function.

File: vj4/cron_jobs/rank.py [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+        {
+          'rankN':rankN,
+          'level':level
+          }})
+    index += 1
+
+  try:
+    bulk.execute()
+  except BulkWriteError as bwe:
+    pprint(bwe.details)
```
Comment (iceboy-sjtu) [inline vj4/cron_jobs/rank.py]: As this is a standalone executable. You should return some non-zero code when error occurs.

File: vj4/cron_jobs/rank.py (hunk 6 above)
Comment (SummerWish) [inline vj4/cron_jobs/rank.py]: Bulk update to improve performance

**Comment by SummerWish** [comment]
@iceboy-sjtu Review

**Pull request merged**